metaflow/plugins/kubernetes/kubernetes_jobsets.py

```diff
@@ -0,0 +1,783 @@
+        "workers_have_started",
+        "all_jobs_are_suspended",
+        "jobset_finished",
+        "jobset_failed",
+        "status_unknown",
+        "jobset_was_terminated",
+        "some_jobs_are_running",
+    ],
+)
+
+
+def _basic_validation_for_js(jobset):
+    if not jobset.get("status") or not _retrieve_replicated_job_statuses(jobset):
+        return False
+    worker_jobs = [
+        w for w in jobset.get("spec").get("replicatedJobs") if w["name"] == "worker"
+    ]
+    if len(worker_jobs) == 0:
+        raise KubernetesJobsetException("No worker jobs found in the jobset manifest")
+    control_job = [
+        w for w in jobset.get("spec").get("replicatedJobs") if w["name"] == "control"
+    ]
+    if len(control_job) == 0:
+        raise KubernetesJobsetException("No control job found in the jobset manifest")
+    return True
+
+
+def _derive_pod_status_and_status_code(control_pod):
+    overall_status = None
+    control_exit_code = None
+    control_pod_failed = False
+    if control_pod:
+        container_status = None
+        pod_status = control_pod.get("status", {}).get("phase")
+        container_statuses = control_pod.get("status", {}).get("containerStatuses")
+        if container_statuses is None:
+            container_status = ": ".join(
+                filter(
+                    None,
+                    [
+                        control_pod.get("status", {}).get("reason"),
+                        control_pod.get("status", {}).get("message"),
+                    ],
+                )
+            )
+        else:
+            for k, v in container_statuses[0].get("state", {}).items():
+                if v is not None:
+                    control_exit_code = v.get("exit_code")
+                    container_status = ": ".join(
+                        filter(
+                            None,
+                            [v.get("reason"), v.get("message")],
+                        )
+                    )
+        if container_status is None:
+            overall_status = (
+                f"pod status: {pod_status} | container status: {container_status}"
+            )
+        else:
+            overall_status = f"pod status: {pod_status}"
+        if pod_status == "Failed":
+            control_pod_failed = True
+    return overall_status, control_exit_code, control_pod_failed
+
+
+def _retrieve_replicated_job_statuses(jobset):
+    # We needed this abstraction because Jobsets changed thier schema
+    # in version v0.3.0 where `ReplicatedJobsStatus` became `replicatedJobsStatus`
+    # So to handle users having an older version of jobsets, we need to account
+    # for both the schemas.
+    if jobset.get("status", {}).get("replicatedJobsStatus", None):
+        return jobset.get("status").get("replicatedJobsStatus")
+    elif jobset.get("status", {}).get("ReplicatedJobsStatus", None):
+        return jobset.get("status").get("ReplicatedJobsStatus")
+    return None
```

Comment:
Needed to add a conditional block to swallow the complexity for the user when the jobset controller may be of a version < 0.3.0. This is because jobset's schema had changed from v0.3.0. No logical changes to the schema, only additions and renaming of values like `ReplicatedJobsStatus` to `replicatedJobsStatus`